server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTask.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.server.util.timer;
+
+public abstract class TimerTask implements Runnable {
+    private volatile TimerTaskEntry timerTaskEntry;
```

Comment:
Perhaps we can use AtomicReference here and remove the synchronized blocks since we just need to protect concurrent access to timerTaskEntry?

Reply:
There is a side effect as well when we mutate `timerTaskEntry`. We call `timerTaskEntry.remove()`. Is it safe to use an `AtomicReference` in this case?

Reply:
I was thinking of something like below which would allow remove to happen outside of the lock (and hence won't block the other threads) which should be fine because TimeTaskList#remove() already performs remove with a lock on specific entry.

With AtomicReference, you will write this code as:
```
TimeTaskEntry oldtaskEntry = timerTaskEntry.getAndSet(null)
if (oldtaskEntry != null)
    oldtaskEntry.remove()
```

Similarily for setTimerEntry(), the code could we written as:
```
TimeTaskEntry oldtaskEntry = timerTaskEntry.getAndSet(newEntry)
if (oldtaskEntry != null && !oldtaskEntry.equals(newEntry)) {
    oldtaskEntry.remove()
}
```

Nevertheless, I would understand if we want to keep this PR similar to scala code to avoid any logic changes. Optimizations suggested here can be done in separate PRs later.

Reply:
Yeah, that should work. I think that it is preferable to do this separately though. My goal was to have a 1:1 mapping between the scala and the java code in this PR.